File: solutions/sonnet-1/sonnet/java/calc.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {
    private String s;
    private int pos;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder out = new StringBuilder();
        String line;
        while ((line = br.readLine()) != null) {
            if (line.trim().isEmpty()) {
                continue;
            }
            Main m = new Main();
            long result = m.eval(line);
            out.append(result).append('\n');
        }
        System.out.print(out);
    }

    private long eval(String line) {
        this.s = line;
        this.pos = 0;
        long result = parseExpr();
        return result;
    }

    private void skipSpaces() {
        while (pos < s.length() && s.charAt(pos) == ' ') {
            pos++;
        }
    }

    private char peek() {
        skipSpaces();
        if (pos < s.length()) {
            return s.charAt(pos);
        }
        return '\0';
    }

    // expr := term (('+' | '-') term)*
    private long parseExpr() {
        long value = parseTerm();
        while (true) {
            char c = peek();
            if (c == '+') {
                pos++;
                value += parseTerm();
            } else if (c == '-') {
                pos++;
                value -= parseTerm();
            } else {
                break;
            }
        }
        return value;
    }

    // term := factor (('*' | '/') factor)*
    private long parseTerm() {
        long value = parseFactor();
        while (true) {
            char c = peek();
            if (c == '*') {
                pos++;
                value *= parseFactor();
            } else if (c == '/') {
                pos++;
                long rhs = parseFactor();
                value = value / rhs;
            } else {
                break;
            }
        }
        return value;
    }

    // factor := number | '(' expr ')'
    private long parseFactor() {
        char c = peek();
        if (c == '(') {
            pos++;
            long value = parseExpr();
            skipSpaces();
            if (pos < s.length() && s.charAt(pos) == ')') {
                pos++;
            }
            return value;
        }
        return parseNumber();
    }

    private long parseNumber() {
        skipSpaces();
        int start = pos;
        while (pos < s.length() && Character.isDigit(s.charAt(pos))) {
            pos++;
        }
        return Long.parseLong(s.substring(start, pos));
    }
}

File: solutions/sonnet-1/sonnet/typescript/calc.ts
class Parser {
  private s: string;
  private i: number = 0;

  constructor(s: string) {
    this.s = s;
  }

  private skipSpaces(): void {
    while (this.i < this.s.length && this.s[this.i] === " ") this.i++;
  }

  private peek(): string {
    this.skipSpaces();
    return this.i < this.s.length ? this.s[this.i] : "";
  }

  parseExpr(): number {
    let value = this.parseTerm();
    while (true) {
      const c = this.peek();
      if (c === "+" || c === "-") {
        this.i++;
        const rhs = this.parseTerm();
        value = c === "+" ? value + rhs : value - rhs;
      } else {
        break;
      }
    }
    return value;
  }

  private parseTerm(): number {
    let value = this.parseFactor();
    while (true) {
      const c = this.peek();
      if (c === "*" || c === "/") {
        this.i++;
        const rhs = this.parseFactor();
        if (c === "*") {
          value = value * rhs;
        } else {
          value = Math.trunc(value / rhs);
        }
      } else {
        break;
      }
    }
    return value;
  }

  private parseFactor(): number {
    const c = this.peek();
    if (c === "(") {
      this.i++;
      const value = this.parseExpr();
      this.skipSpaces();
      if (this.peek() === ")") this.i++;
      return value;
    }
    return this.parseNumber();
  }

  private parseNumber(): number {
    this.skipSpaces();
    let start = this.i;
    while (this.i < this.s.length && this.s[this.i] >= "0" && this.s[this.i] <= "9") {
      this.i++;
    }
    return parseInt(this.s.slice(start, this.i), 10);
  }
}

function evaluate(line: string): number {
  const parser = new Parser(line);
  return parser.parseExpr();
}

async function main() {
  const chunks: Buffer[] = [];
  for await (const chunk of process.stdin) {
    chunks.push(chunk as Buffer);
  }
  const input = Buffer.concat(chunks).toString("utf8");
  const lines = input.split(/\r?\n/);
  const output: string[] = [];
  for (const line of lines) {
    if (line.trim() === "") continue;
    output.push(String(evaluate(line)));
  }
  console.log(output.join("\n"));
}

main();
